### PR TITLE
Testing/bitcoin/witness program

### PR DESCRIPTION
**Pull request by amousa11**

### CODE REVIEW

**Comment by howardwu** [comment]
Heads up, build fails, to fix, change import of `wagu_model` to `wagyu_model`

File: bitcoin/src/witness_program.rs [hunk 1]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+
+        use crate::witness_program::{WitnessProgram, WitnessProgramError};
+
+        const VALID_P2SH_P2WPKH_PROGRAM: &str = "0014751e76e8199196d454941c45d1b3a323f1433bd6";
+        const INVALID_P2SH_P2WPKH_PROGRAM_LENGTH: &str = "0014751e76e8199196d454941c45d1b3a323f143";
+
+        #[test]
+        fn from_str_invalid_p2sh_p2wpkh_program_len() {
+            match WitnessProgram::from_str(INVALID_P2SH_P2WPKH_PROGRAM_LENGTH) {
+                Ok(_) => assert_eq!(None, Some("Error, invalid program decoded successfully")),
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: This case should not be reachable

File: bitcoin/src/witness_program.rs [hunk 2]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+            match WitnessProgram::from_str(VALID_P2SH_P2WPKH_PROGRAM) {
+                Ok(witness_program) => {
+                    assert_eq!(0x00, witness_program.version);
+                    assert_eq!(vec![0x75, 0x1e, 0x76, 0xe8, 0x19,
+                                    0x91, 0x96, 0xd4, 0x54, 0x94,
+                                    0x1c, 0x45, 0xd1, 0xb3, 0xa3,
+                                    0x23, 0xf1, 0x43, 0x3b, 0xd6
+                    ], witness_program.program);
+                }
+                Err(e) => assert_eq!(None, Some(e))
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: This case should not be reachable

File: bitcoin/src/witness_program.rs [hunk 3]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+            }
+        }
+
+        const VALID_P2SH_P2WSH_PROGRAM: &str = "00201863143c14c5166804bd19203356da136c985678cd4d27a1b8c6329604903262";
+        const INVALID_P2SH_P2WSH_PROGRAM_LENGTH: &str = "00201863143c14c5166804bd19203356da136c985678cd4d27a1b8c632960490";
+
+        #[test]
+        fn from_str_invalid_p2sh_p2wsh_program_len() {
+            match WitnessProgram::from_str(INVALID_P2SH_P2WSH_PROGRAM_LENGTH) {
+                Ok(_) => assert_eq!(None, Some("Error, invalid program decoded successfully")),
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: This case should not be reachable

File: bitcoin/src/witness_program.rs [hunk 4]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+                    assert_eq!(vec![0x18, 0x63, 0x14, 0x3c, 0x14,
+                                    0xc5, 0x16, 0x68, 0x04, 0xbd,
+                                    0x19, 0x20, 0x33, 0x56, 0xda,
+                                    0x13, 0x6c, 0x98, 0x56, 0x78,
+                                    0xcd, 0x4d, 0x27, 0xa1, 0xb8,
+                                    0xc6, 0x32, 0x96, 0x04, 0x90,
+                                    0x32, 0x62
+                    ], witness_program.program);
+                }
+                Err(e) => assert_eq!(None, Some(e))
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: This case should not be reachable

File: bitcoin/src/witness_program.rs [hunk 5]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+            }
+        }
+
+        const VALID_OP_1_PROGRAM: &str = "0128751e76e8199196d454941c45d1b3a323f1433bd6751e76e8199196d454941c45d1b3a323f1433bd6";
+        const INVALID_OP_1_PROGRAM_LENGTH: &str = "0128751e76e8199196d454941c45d1b3a323f1433bd6751e76e8199196d454941c45d1b3a323f143";
+
+        #[test]
+        fn from_str_invalid_op_1_program_len() {
+            match WitnessProgram::from_str(INVALID_OP_1_PROGRAM_LENGTH) {
+                Ok(_) => assert_eq!(None, Some("Error, invalid program decoded successfully")),
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: This case should not be reachable

File: bitcoin/src/witness_program.rs [hunk 6]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+                                    0x91, 0x96, 0xd4, 0x54, 0x94,
+                                    0x1c, 0x45, 0xd1, 0xb3, 0xa3,
+                                    0x23, 0xf1, 0x43, 0x3b, 0xd6,
+                                    0x75, 0x1e, 0x76, 0xe8, 0x19,
+                                    0x91, 0x96, 0xd4, 0x54, 0x94,
+                                    0x1c, 0x45, 0xd1, 0xb3, 0xa3,
+                                    0x23, 0xf1, 0x43, 0x3b, 0xd6
+                    ], witness_program.program);
+                }
+                Err(e) => assert_eq!(None, Some(e))
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: This case should not be reachable

File: bitcoin/src/witness_program.rs [hunk 7]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+            }
+        }
+
+        #[test]
+        fn from_str_invalid_hex_str() {
+            match WitnessProgram::from_str("001122zzxxyy") {
+                Ok(_) => assert_eq!(None, Some("Error, invalid hex string decoded successfully")),
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: This case should not be reachable

File: bitcoin/src/witness_program.rs [hunk 8]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+
+    mod test_new {
+        use crate::witness_program::{WitnessProgram, WitnessProgramError};
+
+        const INVALID_VERSION_PROGRAM: &[u8] = &[0x19, 0x03, 0x00, 0x00, 0x00];
+
+        #[test]
+        fn new_invalid_version() {
+            match WitnessProgram::new(INVALID_VERSION_PROGRAM) {
+                Ok(_) => assert_eq!(None, Some("Error, program created with invalid version")),
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: This case should not be reachable

File: bitcoin/src/witness_program.rs [hunk 9]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+                Err(e) => assert_eq!(WitnessProgramError::InvalidVersion(0x19), e)
+            }
+        }
+
+        const INVALID_LENGTH_PROGRAM: &[u8] = &[0x19];
+
+        #[test]
+        fn new_invalid_length() {
+            match WitnessProgram::new(INVALID_LENGTH_PROGRAM) {
+                Ok(_) => assert_eq!(None, Some("Error, program created with invalid version")),
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: This case should not be reachable

File: bitcoin/src/witness_program.rs [hunk 10]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+
+        const INVALID_LENGTH_FOR_VERSION: &[u8] = &[0x00, 0x0f, // Version 0, data length is incorrect
+            0x75, 0x1e, 0x76, 0xe8, 0x19,
+            0x91, 0x96, 0xd4, 0x54, 0x94,
+            0x1c, 0x45, 0xd1, 0xb3, 0xa3];
+
+        #[test]
+        fn new_invalid_program_length_for_version() {
+            match WitnessProgram::new(INVALID_LENGTH_FOR_VERSION) {
+                Ok(_) => assert_eq!(None, Some("Error, program created with invalid version")),
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: This case should not be reachable

File: bitcoin/src/witness_program.rs [hunk 11]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+            0x75, 0x1e, 0x76, 0xe8, 0x19,
+            0x91, 0x96, 0xd4, 0x54, 0x94,
+            0x1c, 0x45, 0xd1, 0xb3, 0xa3,
+            0x23, 0xf1, 0x43, 0x3b, 0xd6,
+            0x00];
+
+        #[test]
+        fn new_invalid_program_length_too_long() {
+            match WitnessProgram::new(INVALID_LENGTH_PROGRAM_TOO_LONG) {
+                Ok(_) => assert_eq!(None, Some("Error, program created with invalid version")),
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: This case should not be reachable

File: bitcoin/src/witness_program.rs (hunk 1 above)
Comment (howardwu) [reply bitcoin/src/witness_program.rs]: See how to test for errors here:
https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L845&L875

File: bitcoin/src/witness_program.rs (hunk 3 above)
Comment (howardwu) [reply bitcoin/src/witness_program.rs]: See how to test for errors here:
https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L845&L875

File: bitcoin/src/witness_program.rs (hunk 5 above)
Comment (howardwu) [reply bitcoin/src/witness_program.rs]: See how to test for errors here:
https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L845&L875

File: bitcoin/src/witness_program.rs (hunk 7 above)
Comment (howardwu) [reply bitcoin/src/witness_program.rs]: See how to test for errors here:
https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L845&L875

File: bitcoin/src/witness_program.rs (hunk 8 above)
Comment (howardwu) [reply bitcoin/src/witness_program.rs]: See how to test for errors here:
https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L845&L875

File: bitcoin/src/witness_program.rs (hunk 9 above)
Comment (howardwu) [reply bitcoin/src/witness_program.rs]: See how to test for errors here:
https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L845&L875

File: bitcoin/src/witness_program.rs (hunk 10 above)
Comment (howardwu) [reply bitcoin/src/witness_program.rs]: See how to test for errors here:
https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L845&L875

File: bitcoin/src/witness_program.rs (hunk 11 above)
Comment (howardwu) [reply bitcoin/src/witness_program.rs]: See how to test for errors here:
https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L845&L875

File: bitcoin/src/witness_program.rs [hunk 12]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
     }
 }
 
-impl fmt::Display for WitnessProgram {
-    /// Prints a witness program in hex representation.
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}", hex::encode(self.to_vec()))
+#[cfg(test)]
+mod tests {
+    mod test_from_str {
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: A general comment, to keep with convention in wagyu, the inner mod should denote the environment the witness program operates in, e.g. `mod p2sh_p2wpkh_mainnet`.

See https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L676

File: bitcoin/src/witness_program.rs [hunk 13]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+        #[test]
+        fn from_str_invalid_hex_str() {
+            match WitnessProgram::from_str("001122zzxxyy") {
+                Ok(_) => assert_eq!(None, Some("Error, invalid hex string decoded successfully")),
+                Err(e) => assert_eq!(WitnessProgramError::ProgramDecodingError, e)
+            }
+        }
+    }
+
+    mod test_new {
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: A general comment, to keep with convention in wagyu, the inner mod should denote the environment the witness program operates in, e.g. `mod p2sh_p2wpkh_mainnet`.

See https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L676

File: bitcoin/src/witness_program.rs [hunk 14]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+        #[test]
+        fn new_invalid_program_length_too_long() {
+            match WitnessProgram::new(INVALID_LENGTH_PROGRAM_TOO_LONG) {
+                Ok(_) => assert_eq!(None, Some("Error, program created with invalid version")),
+                Err(e) => assert_eq!(WitnessProgramError::InvalidProgramLength(41), e)
+            }
+        }
+    }
+
+    mod test_to_scriptpubkey {
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: A general comment, to keep with convention in wagyu, the inner mod should denote the environment the witness program operates in, e.g. `mod p2sh_p2wpkh_mainnet`.

See https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L676

File: bitcoin/src/witness_program.rs [hunk 15]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}", hex::encode(self.to_vec()))
+#[cfg(test)]
+mod tests {
+    mod test_from_str {
+        use std::str::FromStr;
+
+        use crate::witness_program::{WitnessProgram, WitnessProgramError};
+
+        const VALID_P2SH_P2WPKH_PROGRAM: &str = "0014751e76e8199196d454941c45d1b3a323f1433bd6";
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Adhere to the test suite convention of providing a slice containing all programs at the top of the inner mod.

See https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L682

File: bitcoin/src/witness_program.rs [hunk 16]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+                                    0x91, 0x96, 0xd4, 0x54, 0x94,
+                                    0x1c, 0x45, 0xd1, 0xb3, 0xa3,
+                                    0x23, 0xf1, 0x43, 0x3b, 0xd6
+                    ], witness_program.program);
+                }
+                Err(e) => assert_eq!(None, Some(e))
+            }
+        }
+
+        const VALID_P2SH_P2WSH_PROGRAM: &str = "00201863143c14c5166804bd19203356da136c985678cd4d27a1b8c6329604903262";
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Adhere to the test suite convention of providing a slice containing all programs at the top of the inner mod.

See https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L682

File: bitcoin/src/witness_program.rs [hunk 17]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+                }
+            }
+        }
+
+        #[test]
+        fn from_str_valid_p2sh_p2wsh_program() {
+            match WitnessProgram::from_str(VALID_P2SH_P2WSH_PROGRAM) {
+                Ok(witness_program) => {
+                    assert_eq!(0x00, witness_program.version);
+                    assert_eq!(vec![0x18, 0x63, 0x14, 0x3c, 0x14,
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Adhere to the test suite convention of providing a slice containing all programs at the top of the inner mod.

See https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L682

File: bitcoin/src/witness_program.rs [hunk 18]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+                                    0xcd, 0x4d, 0x27, 0xa1, 0xb8,
+                                    0xc6, 0x32, 0x96, 0x04, 0x90,
+                                    0x32, 0x62
+                    ], witness_program.program);
+                }
+                Err(e) => assert_eq!(None, Some(e))
+            }
+        }
+
+        const VALID_OP_1_PROGRAM: &str = "0128751e76e8199196d454941c45d1b3a323f1433bd6751e76e8199196d454941c45d1b3a323f1433bd6";
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Adhere to the test suite convention of providing a slice containing all programs at the top of the inner mod.

See https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L682

File: bitcoin/src/witness_program.rs [hunk 19]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+                }
+            }
+        }
+
+        #[test]
+        fn from_str_valid_op_1_program() {
+            match WitnessProgram::from_str(VALID_OP_1_PROGRAM) {
+                Ok(witness_program) => {
+                    assert_eq!(0x01, witness_program.version);
+                    assert_eq!(vec![0x75, 0x1e, 0x76, 0xe8, 0x19,
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Adhere to the test suite convention of providing a slice containing all programs at the top of the inner mod.

See https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L682

File: bitcoin/src/witness_program.rs [hunk 20]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+
+        #[test]
+        fn new_invalid_version() {
+            match WitnessProgram::new(INVALID_VERSION_PROGRAM) {
+                Ok(_) => assert_eq!(None, Some("Error, program created with invalid version")),
+                Err(e) => assert_eq!(WitnessProgramError::InvalidVersion(0x19), e)
+            }
+        }
+
+        const INVALID_LENGTH_PROGRAM: &[u8] = &[0x19];
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Move isolated invalid cases into dedicated functions or declare all invalid const cases up top

File: bitcoin/src/witness_program.rs [hunk 21]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+
+        #[test]
+        fn new_invalid_length() {
+            match WitnessProgram::new(INVALID_LENGTH_PROGRAM) {
+                Ok(_) => assert_eq!(None, Some("Error, program created with invalid version")),
+                Err(e) => assert_eq!(WitnessProgramError::InvalidProgramLength(1), e)
+            }
+        }
+
+        const INVALID_LENGTH_FOR_VERSION: &[u8] = &[0x00, 0x0f, // Version 0, data length is incorrect
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Move isolated invalid cases into dedicated functions or declare all invalid const cases up top

File: bitcoin/src/witness_program.rs [hunk 22]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+
+        #[test]
+        fn new_invalid_program_length_for_version() {
+            match WitnessProgram::new(INVALID_LENGTH_FOR_VERSION) {
+                Ok(_) => assert_eq!(None, Some("Error, program created with invalid version")),
+                Err(e) => assert_eq!(WitnessProgramError::InvalidProgramLengthForVersion(15, 0x00), e)
+            }
+        }
+
+        const INVALID_LENGTH_PROGRAM_TOO_LONG: &[u8] = &[0x00, 0x29,
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Move isolated invalid cases into dedicated functions or declare all invalid const cases up top

File: bitcoin/src/witness_program.rs [hunk 23]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+    }
+
+    mod test_to_scriptpubkey {
+        use crate::witness_program::WitnessProgram;
+
+        #[test]
+        fn to_scriptpubkey_valid_p2sh_p2wpkh_program() {
+            let witness_program = WitnessProgram {
+                version: 0x00,
+                program: vec![
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Adhere to the test suite convention of providing a slice containing all programs at the top of the inner mod.

See https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L682

File: bitcoin/src/witness_program.rs [hunk 24]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+                            0x1c, 0x45, 0xd1, 0xb3, 0xa3,
+                            0x23, 0xf1, 0x43, 0x3b, 0xd6
+            ], witness_program.to_scriptpubkey());
+        }
+
+        #[test]
+        fn to_scriptpubkey_valid_p2sh_p2wsh_program() {
+            let witness_program = WitnessProgram {
+                version: 0x00,
+                program: vec![0x18, 0x63, 0x14, 0x3c, 0x14,
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Adhere to the test suite convention of providing a slice containing all programs at the top of the inner mod.

See https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L682

File: bitcoin/src/witness_program.rs [hunk 25]
```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
...
+                            0xc6, 0x32, 0x96, 0x04, 0x90,
+                            0x32, 0x62
+            ], witness_program.to_scriptpubkey());
+        }
+
+        #[test]
+        fn to_scriptpubkey_valid_op_1_program() {
+            let witness_program = WitnessProgram {
+                version: 0x01,
+                program: vec![0x75, 0x1e, 0x76, 0xe8, 0x19,
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Adhere to the test suite convention of providing a slice containing all programs at the top of the inner mod.

See https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L682

File: bitcoin/src/witness_program.rs [hunk 26]
```diff
@@ -20,29 +20,54 @@
...
+impl From<WitnessProgramError> for AddressError {
+    fn from(error: WitnessProgramError) -> Self {
+        AddressError::Crate("WitnessProgram", format!("{:?}", error))
+    }
+}
+
 impl WitnessProgram {
-    /// Returns a new witness program given a program with a version byte, data size, and data.
-    pub fn new(program: &[u8]) -> Result<WitnessProgram, &'static str> {
+    /// Returns a new witness program given a program with a version byte (where 0 <= version <= 16), data size, and data.
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: `with a version byte (where 0 <= version <= 16)`

Please explain your logic.

File: bitcoin/src/witness_program.rs [hunk 27]
```diff
@@ -20,29 +20,54 @@
...
+#[derive(Debug, Fail, PartialEq, Eq)]
+pub enum WitnessProgramError {
+    #[fail(display = "expected program with length at least 2, at most 40, actual program was length {}", _0)]
+    InvalidProgramLength(usize),
+
+    #[fail(display = "invalid program length {} for script version {}", _0, _1)]
+    InvalidProgramLengthForVersion(usize, u8),
+
+    #[fail(display = "expected version no greater than 16")]
+    InvalidVersion(u8),
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Unused parameter and remove the mention of 16

File: bitcoin/src/witness_program.rs [hunk 28]
```diff
@@ -20,29 +20,54 @@
...
 pub struct WitnessProgram {
     /// The version byte
     pub version: u8,
     /// The witness program bytes
     pub program: Vec<u8>,
 }
 
+#[derive(Debug, Fail, PartialEq, Eq)]
+pub enum WitnessProgramError {
+    #[fail(display = "expected program with length at least 2, at most 40, actual program was length {}", _0)]
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Please condense this description and remove the mention of 2 and 40

File: bitcoin/src/witness_program.rs [hunk 29]
```diff
@@ -20,29 +20,54 @@
...
+    InvalidProgramLength(usize),
+
+    #[fail(display = "invalid program length {} for script version {}", _0, _1)]
+    InvalidProgramLengthForVersion(usize, u8),
+
+    #[fail(display = "expected version no greater than 16")]
+    InvalidVersion(u8),
+
+    #[fail(display = "expected program length {} to equal length {} as specified by input", _0, _1)]
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]:
```suggestion
    #[fail(display = "invalid program length: {{ expected: {:?}, found: {:?} }}", _0, _1)]
```

File: bitcoin/src/witness_program.rs [hunk 30]
```diff
@@ -20,29 +20,54 @@
...
+    MismatchedProgramLength(usize, usize),
+
+    #[fail(display = "error decoding program from hex string")]
+    ProgramDecodingError,
+}
+
+impl From<WitnessProgramError> for AddressError {
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Why is this here? Remove

File: bitcoin/src/witness_program.rs [hunk 31]
```diff
@@ -20,29 +20,54 @@
 //! If the version byte is 0, but the witness program is neither 20 nor 32 bytes, the script must fail.
 //!
 
-use hex;
-use std::fmt;
 use std::str::FromStr;
 
+use wagyu_model::AddressError;
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Remove

File: bitcoin/src/witness_program.rs [hunk 32]
```diff
@@ -20,29 +20,54 @@
...
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Can a witness program be partially ordered, ordered, and hashed? Please explain

File: bitcoin/src/witness_program.rs [hunk 33]
```diff
@@ -20,29 +20,54 @@
...
 pub struct WitnessProgram {
     /// The version byte
     pub version: u8,
     /// The witness program bytes
     pub program: Vec<u8>,
 }
 
+#[derive(Debug, Fail, PartialEq, Eq)]
+pub enum WitnessProgramError {
```
Comment (howardwu) [inline bitcoin/src/witness_program.rs]: Nit: move this above `pub struct WitnessProgram`

File: bitcoin/src/witness_program.rs (hunk 30 above)
Comment (amousa11) [reply bitcoin/src/witness_program.rs]: https://github.com/ArgusHQ/wagyu/blob/d455e031ff1866ff76d255c311d69ab5f7c39e9c/bitcoin/src/address.rs#L220

The `?` on this line requires this

File: bitcoin/src/witness_program.rs (hunk 26 above)
Comment (amousa11) [reply bitcoin/src/witness_program.rs]: for a witness program, the first byte specifies the witness program version. The version is a value between 0 and 16 inclusive.

File: bitcoin/src/witness_program.rs (hunk 30 above)
Comment (howardwu) [reply bitcoin/src/witness_program.rs]: This should be under `model/address.rs` as it's an `impl` of `AddressError`. You'll find all of its cousin `{Model}Error` conversions there as well.

File: bitcoin/src/witness_program.rs (hunk 26 above)
Comment (howardwu) [reply bitcoin/src/witness_program.rs]: A byte is 8 bits, (2^8)-1 == 255

File: bitcoin/src/witness_program.rs (hunk 30 above)
Comment (amousa11) [reply bitcoin/src/witness_program.rs]: resolved offline
